src/ott/neural/solvers/expectile_neural_dual.py

```diff
@@ -0,0 +1,655 @@
+
+  def transport(self, vec: jnp.ndarray, forward: bool = True) -> jnp.ndarray:
+    """Transport from source to the target distribution or vice-versa."""
+    vec = jnp.atleast_2d(vec)
+    if forward:
+      return vec - self._grad_h_inv(self._grad_f(vec))
+    return vec - self._grad_h_inv(self._grad_g(vec))
+
+
+class ForwardPotentials(DualPotentials):
+  r"""The dual potentials for only forward mapping.
+
+  Args:
+  g: The second dual potential function.
+  grad_f: Gradient of the first dual potential function.
+  cost_fn: The cost function used to solve the OT problem.
+  """
+
+  def __init__(
+      self, g: PotentialValueFn_t, grad_f: PotentialGradientFn_t,
+      cost_fn: costs.CostFn
+  ):
+
+    def g_cost_conjugate(x: jnp.ndarray) -> jnp.ndarray:
+      y_hat = jax.lax.stop_gradient(grad_f(x))
+      return -g(y_hat) + cost_fn(x, y_hat)
+
+    super().__init__(g_cost_conjugate, g, cost_fn=cost_fn, corr=False)
+    self.__grad_f = grad_f
+
+  @property
+  def _grad_f(self) -> Callable[[jnp.ndarray], jnp.ndarray]:
+    return jax.vmap(self.__grad_f)
+
+  def transport(self, source: jnp.ndarray, forward: bool = True) -> jnp.ndarray:
+    """Transport the points from source to the target distribution."""
+    source = jnp.atleast_2d(source)
+    assert (forward
+            is True), ("Only forward mapping (source -> target) is supported.")
+    return self._grad_f(source)
+
+
+class EuclideanPotentials(DualPotentials):
+  r"""The dual potentials, trained with scalar product operation.
+
+  Args:
+  g: The second dual potential function.
+  grad_f: Gradient of the first dual potential function.
+  """
+
+  def __init__(self, g: PotentialValueFn_t, grad_f: PotentialGradientFn_t):
+
+    def g_conjugate(x: jnp.ndarray) -> jnp.ndarray:
+      y_hat = jax.lax.stop_gradient(grad_f(x))
+      return -g(y_hat) + jnp.dot(x, y_hat)
+
+    super().__init__(g_conjugate, g, cost_fn=costs.SqEuclidean(), corr=True)
+    self.__grad_f = grad_f
+
+  @property
+  def _grad_f(self) -> Callable[[jnp.ndarray], jnp.ndarray]:
+    return jax.vmap(self.__grad_f)
+
+  def transport(self, vec: jnp.ndarray, forward: bool = True) -> jnp.ndarray:
+    """Transport from source to the target distribution or vice-versa."""
+    vec = jnp.atleast_2d(vec)
+    return self._grad_f(vec) if forward else self._grad_g(vec)
+
+  def distance(self, src: jnp.ndarray, tgt: jnp.ndarray) -> float:
+    """W2 distance."""
+    src, tgt = jnp.atleast_2d(src), jnp.atleast_2d(tgt)
+    f = jax.vmap(self.f)
+    g = jax.vmap(self.g)
+    corr = jnp.mean(f(src)) + jnp.mean(g(tgt))
+    return -2.0 * corr + jnp.mean(jnp.sum(src ** 2, axis=-1)) + \
+      jnp.mean(jnp.sum(tgt ** 2, axis=-1))
+
+
+def expectile_loss(
+    adv: jnp.ndarray, diff: jnp.ndarray, expectile: float = 0.9
+) -> jnp.ndarray:
+  weight = jnp.where(adv >= 0, expectile, (1 - expectile))
+  return weight * diff ** 2
+
+
+@jax.tree_util.register_pytree_node_class
+class DotCost(costs.CostFn):
```

Comment:
if this is needed, maybe better add it in `costs.py`

Reply:
If it's useful somewhere other than this solver, I can move it to costs.py.

Reply:
Class DotCost removed. Replaced by `lambda x, y: -jax.vmap(jnp.dot)(x, y)`